### ISSUE TITLE
Java14 Full Records Support Check Validation: InterfaceMemberImpliedModifierCheck

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8508]
Child of #8452 
Check documentation: ~~https://checkstyle.sourceforge.io/config_modifier.html#ClassMemberImpliedModifier~~
https://checkstyle.sourceforge.io/config_modifier.html#InterfaceMemberImpliedModifier

From check documentation:
_Checks for implicit modifiers on interface members and nested types._

_This check is effectively the opposite of RedundantModifier. It checks the modifiers on interface members, ensuring that certain modifiers are explicitly specified even though they are actually redundant._

```
➜  full-record-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  full-record-grammar cat config.xml                                                                    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="InterfaceMemberImpliedModifier"/>
  </module>
</module>
➜  full-record-grammar cat TestClass.java                                                                
class TestClass {
    public interface RoadFeature {

        String STOP = "Stop";  // violation x3

        final public static String STOP2 = "Stop2";  // ok

        enum Lights {  // violation x2

            RED, YELLOW, GREEN;
        }
    }
}

record MyRecord1(boolean t, boolean f) {
    public interface RoadFeature {

        String STOP = "Stop";  // violation x3

        final public static String STOP2 = "Stop2";  // ok

        enum Lights {  // violation x2

            RED, YELLOW, GREEN;
        }
    }
}







➜  full-record-grammar java $RUN_LOCALE -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.35-SNAPSHOT-all.jar -c config.xml TestClass.java
Starting audit...
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:4:9: Implied modifier 'final' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:4:9: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:4:9: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:8:9: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:8:9: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:18:9: Implied modifier 'final' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:18:9: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:18:9: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:22:9: Implied modifier 'public' should be explicit. [InterfaceMemberImpliedModifier]
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:22:9: Implied modifier 'static' should be explicit. [InterfaceMemberImpliedModifier]
Audit done.
Checkstyle ends with 10 errors.

```

This check works as intended within the body of a record definition, so I think we can close this one.

**Comment by nmancus1** [comment]
@esilkensen @pbludov ping

**Comment by pbludov** [comment]
Sorry, i've missed this issue.

This check should be updated. The `records` are always `static` as `enums` and `interfaces`:
```java
public class TestClass {
    int field;
    public record x() {
        public x {
            System.out.println(field); // Non-static variable field cannot be referenced from a static context
        }
    }
}
```
The test case should be
```java
public class TestClass {
    public static interface GoodInterface {} // OK
    public interface BadInterface {} // Implied modifier 'static' should be explicit. [ClassMemberImpliedModifier]

    public static enum GoodEnum {} // OK
    public enum BadEnum {} // Implied modifier 'static' should be explicit. [ClassMemberImpliedModifier]

    public static record GoodRecord() {} // OK
    public record BadRecord() {} // expected same violation
}
```
The check should be updated to support `TokenTypes.RECPRD_DEF`.
Also, a new property `violateImpliedStaticOnNestedRecord` (`true` by default) should be added.

**Comment by nmancus1** [comment]
@pbludov I realized that I made a mistake on this issue; I linked the documentation for ClassMemberImpliedModifierCheck in the issue statement.  From what you've shared above, I take it that you meant for that test input to be for ClassMemberImpliedModifier and not InterfaceMemberImpliedModifierCheck.  Is that correct?  I have fixed the link to the documentation.
